lowranceusr.cc

```diff
@@ -455,226 +1378,361 @@ lowranceusr_parse_routes()
+          /* option to break trails into segments was specified */
+          auto trk_tmp = route_head_alloc();
+          trk_tmp->rte_num = ++(*trail_num);
+          trk_tmp->rte_name = trk_head->rte_name;
+          track_add_head(trk_tmp);
+          trk_head = trk_tmp;
+        }
+        track_add_wpt(trk_head, wpt_tmp);
+
+        if (global_opts.debug_level > 2) {
+          printf(MYNAME " parse_trails: Trail pt lat %f lon %f\n", wpt_tmp->latitude, wpt_tmp->longitude);
+        }
+      }
+    }
+  } else {
+    /* remove the trail since it's empty */
+    track_del_head(trk_head);
+  }
 }
 
-/*
- * Icons are automatically converted to waypoints unless
- * option of ignoreicons is used
- */
 static void
-lowranceusr_parse_icons()
+lowranceusr4_parse_trail(int* trail_num)
 {
-  char buff[MAXUSRSTRINGSIZE + 1];
-
-  short int num_icons = gbfgetint16(file_in);
+  int trail_version;
+  char name_buff[MAXUSRSTRINGSIZE + 1];
+  char desc_buff[MAXUSRSTRINGSIZE + 1];
+  int trail_color;
+  int trail_flags;
+
+  lowranceusr4_fsdata* fsdata = lowranceusr4_alloc_fsdata();
+  fs_chain_add(&(trk_head->fs), (format_specific_data*) fsdata);
+
+  /* UID unit number */
+  fsdata->uid_unit = gbfgetint32(file_in);
+
+  /* 64-bit UID sequence number */
+  fsdata->uid_seq_low = gbfgetint32(file_in);
+  fsdata->uid_seq_high = gbfgetint32(file_in);
+
+  /* Trail stream version number */
+  trail_version = gbfgetint16(file_in);
+  if (global_opts.debug_level == 99) {
+    printf(MYNAME " parse_trails: trail Version %d\n", trail_version);
+  }
 
-  if (global_opts.debug_level >= 1) {
-    printf(MYNAME " parse_icons: num Icons = %d\n", num_icons);
+  /* Trail name; input is 2 bytes per char, we convert to 1 */
+  int name_len = lowranceusr4_readstr(&name_buff[0], MAXUSRSTRINGSIZE, file_in, 2);
+  if (name_len) {
+    name_buff[name_len] = '\0';
+    trk_head->rte_name = name_buff;
   }
 
-  for (int i = 0; i < num_icons && !gbfeof(file_in); i++) {
-    if (ignoreicons) {
-      /* position coord lat & long */
-      (void) gbfread(&buff[0], 4, 2, file_in);
-      /* symbol */
-      (void) gbfread(&buff[0], 4, 1, file_in);
-    } else {
-      Waypoint* wpt_tmp = new Waypoint;
+  /* Flags, discard for now */
+  trail_flags = gbfgetint32(file_in);
 
-      /* position coord lat & long */
-      wpt_tmp->latitude = lat_mm_to_deg(gbfgetint32(file_in));
-      wpt_tmp->longitude = lon_mm_to_deg(gbfgetint32(file_in));
-      wpt_tmp->altitude = 0;
-      snprintf(buff, sizeof(buff), "Icon %d", i+1);
-      wpt_tmp->shortname = buff;
-      /* symbol */
-      wpt_tmp->icon_descr = lowranceusr_find_desc_from_icon_number(gbfgetint32(file_in));
-      waypt_add(wpt_tmp);
-    }
+  /* Color ID, discard for now */
+  trail_color = gbfgetint32(file_in);
+
+  /* Comment/description */
+  int desc_len = lowranceusr4_readstr(&desc_buff[0], MAXUSRSTRINGSIZE, file_in, 2);
+  if (desc_len) {
+    desc_buff[desc_len] = '\0';
+    trk_head->rte_desc = desc_buff;
```

Comment:
WIth foo_readstr() returning QStrings, we'd be out of the business of manually terminating and testing them before we used them.About every caller would benefit.

Reply:
I already have a patch that does this.